src/codec/syslog.rs

```diff
@@ -0,0 +1,460 @@
+        );
+        es_msg.insert(
+            "protocol".into(),
+            Value::from(match parsed.protocol {
+                Protocol::RFC3164 => "RFC3164",
+                Protocol::RFC5424(_) => "RFC5424",
+            }),
+        );
+        es_msg.insert(
+            "protocol_version".into(),
+            match parsed.protocol {
+                Protocol::RFC5424(version) => Value::from(version),
+                Protocol::RFC3164 => Value::null(),
+            },
+        );
+        es_msg.insert("appname".into(), Value::from(parsed.appname));
+        es_msg.insert("msgid".into(), Value::from(parsed.msgid));
+
+        let mut temp = Object::with_capacity(parsed.structured_data.len());
+        for element in parsed.structured_data {
+            let mut e = Vec::with_capacity(element.params.len());
+            for (name, value) in element.params {
+                let mut param = Value::object_with_capacity(1);
+                param.insert(name.to_owned(), Value::from(value))?;
+                e.push(param);
+            }
+            temp.insert(element.id.into(), Value::Array(e));
+        }
+        if temp.is_empty() {
+            es_msg.insert("structured_data".into(), Value::null());
+        } else {
+            es_msg.insert("structured_data".into(), Value::from(temp));
+        }
+
+        if let Some(procid) = parsed.procid {
+            let value: Value = match procid {
+                ProcId::PID(pid) => pid.to_string().into(),
+                ProcId::Name(name) => name.to_string().into(),
+            };
+            es_msg.insert("procid".into(), value);
+        } else {
+            es_msg.insert("procid".into(), Value::null());
+        }
+
+        es_msg.insert("msg".into(), Value::from(parsed.msg));
+
+        Ok(Some(Value::from(es_msg)))
+    }
+
+    #[allow(clippy::too_many_lines)]
+    fn encode(&self, data: &Value) -> Result<Vec<u8>> {
+        let mut result: Vec<String> = Vec::new();
+
+        let mut v = String::new();
+        let protocol = match (data.get_str("protocol"), data.get_u32("protocol_version")) {
+            (Some("RFC3164"), _) => Protocol::RFC3164,
+            (Some("RFC5424"), Some(version)) => {
+                v = version.to_owned().to_string();
+                Protocol::RFC5424(version)
+            }
+            (Some("RFC5424"), None) => {
+                return Err(ErrorKind::InvalidSyslogData("Missing protocol version".into()).into())
+            }
+            (None, Some(_)) => {
+                return Err(ErrorKind::InvalidSyslogData("Missing protocol type".into()).into())
+            }
+            (Some(&_), _) => {
+                return Err(ErrorKind::InvalidSyslogData("invalid protocol type".into()).into())
+            }
+            (None, None) => {
+                v = "1".to_owned();
+                Protocol::RFC5424(1_u32)
+            }
+        };
+
+        if let Some(f) = data.get_str("facility") {
```

Comment:
th `if let ... else { return Error... }` pattern can in many places be simplified / easier to handle when using [`Option::ok_or_else`]:

```rust
if let Some(f) = data.get_str("facility") {
 // ...
} else {
 return Err(ErrorKind::InvalidSyslogData("facility missing".into()).into());
}
```

can be rewritten as:

```rust
let f = data.get_str("facility").ok_or_else(|| Error::from(ErrorKind::InvalidSyslogData("facility missing".into())))?;
 // ...
```

it's not always the right approach bit it often makes the code more readable as it removes nested `if`'s